### ISSUE TITLE
Patterns and related features

### ISSUE THREAD

**Issue by munificent** [dart-lang/language#546]
We're starting to investigate a couple of related language changes around tuples, pattern matching, and destructuring. These features touch on other things like "data classes", so I wanted to kick things off by trying to sketch out my understanding of the space, how I think the various features relate to each other, and the relevant issues people have filed.

Draft feature specification:
https://github.com/dart-lang/language/blob/master/accepted/future-releases/0546-patterns/feature-specification.md

Implementation issue: https://github.com/dart-lang/sdk/issues/50585

## Concepts

As I understand it, the core concepts are:

### Patterns

Languages with [pattern matching](https://en.wikipedia.org/wiki/Pattern_matching#Tree_patterns) revolve around **patterns**. "Expression" and "statement" are both syntactic categories in the grammar. Patterns form a third category. Most languages that have pattern matching have a variety of different kinds of patterns they support. The basic idea with a pattern is that it:

* Can be tested against some value to determine if the pattern *matches*. Some kinds of patterns always match.
* If (and only if) it does match, the pattern may bind a new variable in some scope.

Languages with patterns use them in a variety of places. They can be how you declare variables and catch exceptions. Some languages use them for defining parameter lists or "overloads" of a function. Every language with patterns has some kind of explicit pattern matching expression or statement...

* https://github.com/dart-lang/language/issues/112 (pattern for catch clauses)
* https://github.com/dart-lang/language/issues/148 (case functions)

### Pattern matching

Once you have patterns, it makes sense to introduce a control flow structure that uses them. A **pattern matching statement** takes a value that it matches against. Then it has a series of pairs of patterns and bodies. At runtime, the implementation tests each pattern against the value in turn. The first pattern that matches has its body executed. Any variables the pattern binds are only available in the corresponding body.

If you really want a functional style, an even more powerful form is a **pattern matching expression** that lets you do pattern matching in the middle of a larger expression and have it evaluate to a value. In order to do that in a sound way, though, you need the next concept...

* https://github.com/dart-lang/language/issues/27 (switch expression)
* https://github.com/dart-lang/language/issues/307 (switch expression)

### Exhaustiveness

So you're executing a pattern matching construct, walking through all of the patterns to find a match. What happens if *none* of the cases match? If the pattern matching construct is an *expression*, this is a real problem because what value do you evaluate to in that case? Even when you are using pattern matching as a statement, users still likely want to know if it's possible for a value to not match any of the cases.

To help with that, many languages with pattern matching also do **exhaustiveness checking**. The static checker will examine all of the patterns and determine if it's possible for a value to sneak through and match none of them. If so, the compiler reports a warning or error to let the user know. Dart does a limited form of this now with switch statements on enum types. If you miss one of the enum cases, Dart gives you a warning to let you know. Exhaustiveness checking in pattern matching takes that concept and applies it to much larger, more complex patterns.

### Destructuring

A list literal expression takes a number of smaller values, the list element expressions, and composes a new larger value out of them, the list. Patterns go in the other direction. For example, a list pattern contains a series of nested patterns for elements. When the list pattern is matched against a list, it **destructures** the list by walking the elements and matching them against the corresponding element patterns.

This gives you a really nice terse way to pull pieces of aggregate objects like lists, maps, and even instances of classes. Any time you have a pattern that contains nested subpatterns, you're usually doing some kind of destructuring.

* https://github.com/dart-lang/language/issues/207 (destructuring)

### Tuples

Lists are a great way to collect a series of values of the *same* type, but they work poorly when you want to, say return a number and a string from a function. Since lists only have a single element type, the type of each separate element is lost.

Tuples fix that. A tuple is sort of like a fixed-size list where each element has its own distinct type. The type system can see "into" the tuple. A tuple of an int followed by a String has a different static type than a tuple of two booleans. Syntactically, tuples are usually a comma-separated list of expressions surrounded by parentheses, like `(1, "string")`.

Tuples are useful in a statically-typed language for things like **multiple return values** because they maintain precise types. Once you have a tuple, you eventually need to pull the elements back out of it. You can provide number-like getters, but that gets tedious. Languages with tuples almost always have some kind of **tuple pattern** so that you can destructure them.

* https://github.com/dart-lang/language/issues/68 (multiple return values)

### Algebraic data types, sum types, unions with fields

Pattern matching comes to us from the ML language family. Another key feature of those languages that ties into pattern matching are **[algebraic data types][]**, also known as **sum types**, **discriminated unions**, or **tagged unions**. To make matters more confusing, these are quite different from both **[union types][]** and (untagged) **unions** in C and C++. Algebraic data types are sometimes abbreviated **ADT**, which is again distinct from **abstract data types**. Thanks, computer scientists.

[algebraic data types]: https://en.wikipedia.org/wiki/Algebraic_data_type
[union types]: https://dotty.epfl.ch/docs/reference/new-types/union-types.html

Sum types are like superpowered unions. You have a type that contains a fixed, closed set of cases. Unlike Dart enums, each case may also have its own fields containing extra data. For example, you might have a Shape type, with cases for Rect and Circle. Rect would have four coordinates for its corners. Circle would have a center and radius.

In an object-oriented language like Dart, you'd model this using subclasses. You'd have an abstract Shape class and Rect and Circle subclasses. Subclasses and ADTs are sort of duals. Indeed, Scala's [case classes][] are like ADTs but are subclasses under the hood.

[case classes]: https://docs.scala-lang.org/tour/case-classes.html

These come into discussions of pattern matching because the typical way to define behavior on specific cases in a sum type is by using a pattern match on the different cases. Likewise, the way to extract each case's fields is by using a case pattern to destructure them. Something like:

```scala
someShape match {
  case Rect(left, top, right, bottom) => ...
  case Circle(x, y, radius) => ...
}
```

It would be very strange to add algebraic data types to a language without a nice way to switch on and decompose them like this.

* https://github.com/dart-lang/language/issues/83 ("sum type", though the request is more for union types)
* https://github.com/dart-lang/language/issues/145 (union types)
* https://github.com/dart-lang/language/issues/158 (fields in enums)
* https://github.com/dart-lang/language/issues/183 (more powerful enums)
* https://github.com/dart-lang/language/issues/349 (algebraic data types)

### "Data classes" and "value types"

One of the most frequent feature requests for Dart is some kind of **data classes** or **value types**. The former is inspired by [Kotlin's corresponding feature][data class]. The latter means different things to different people, but common attributes seem to be:

* A lightweight way of defining a named type with some fields without having to explicitly declare a constructor that initializes each field.

* An automatically provided implementation of `hashCode` and `operator ==()` so that the resulting object has "value semantics".

* Some kind of **immutability**. The fields can't be modified. Some ask for **deep immutability**—the values of all fields must themselves be immutable. This usually implies that the data class can't be extended either.

* Often some easy way to clone or make a new copy of an object with some fields changed.

[data class]: https://kotlinlang.org/docs/reference/data-classes.html

Data classes are involved with patterns and pattern matching because users also often request that a data class also implicitly provides destructuring support. (Kotlin data classes give you this.) That means some kind of pattern to let you pull apart the fields in an object.

Also, part of the motivation for both algebraic data types and data classes is a nicer notation for defining a new composite type without all of the boilerplate overhead of a full class declaration. In other words, a good ADT feature might subsume data classes or vice versa.

* https://github.com/dart-lang/language/issues/125 (immutable objects)
* https://github.com/dart-lang/language/issues/138 (primary constructor)
* https://github.com/dart-lang/language/issues/314 (data classes)

### Type patterns

The last concept which has mostly come up internally but exposes a capability the language doesn't otherwise offer is some kind of way to expose the *runtime* type argument of an object. If you have patterns and pattern matching, a natural solution is a **type pattern** that matches an object of some generic type and binds variables to the object's type arguments.

* https://github.com/dart-lang/language/issues/169 (type decomposition)
* https://github.com/dart-lang/language/issues/170 (type patterns)

## Structure

We are *very* early in our investigation. This is a family of features that are very close to my heart. I wrote a doc in *2011* requesting that we add pattern matching to Dart. At the same time, as you can see, this is a big sprawling feature and we currently have our hands full with extension methods and non-nullable types (both *very* large features in their own right!).

That being said, here is how I am interested in approaching the space and what I hope we can do:

1. **Define a set of pattern syntax and semantics.** I think patterns are the most important core to all of these features. You obviously can't do pattern matching, destructuring, and exhaustiveness checking at all without them. You technically *can* do tuples, sum types, and data classes, but I think they are significantly hampered without them.

    Also, while it may not be obvious, I believe there are some real tricky design challenges in this area. Patterns are a dense grammatical region where we want to be able to express a wide variety of behaviors in a small amount of syntax. We need to be able to destructure lists, maps, sets, and tuples. Do runtime type tests. Pull fields out of, at least, sum types or data classes. And, of course, test for literals and constant values for simple switch-like cases.

    We are constrained by the expectations users bring from other languages, the desire for patterns to mirror their corresponding expression forms, and (hardest) Dart's existing notation for variable declarations, since those are already a limited form of "pattern" in that they introduce bindings.

    I think getting this right is key.

2.  **Define a set of language constructs where patterns can be used.** Some new kind of pattern matching statement is the obvious one. But also retrofitting them into variable declarations so that you can destructure in a declaration. Can we re-engineer catch clauses to be pattern based so that you can have more sophisticated catches? Should parameter lists use patterns? Do we want an expression-based pattern match? Some kind of if-let-like statement? This is the fun stuff. Once users have internalized how patterns work, the more places they can use that knowledge the better.

3.  **Add tuples.** This can happen in parallel with the previous one. Much like adding sets, this is a whole feature in its own right with new syntax, object representation, runtime behavior, and type checking.

    @lrhn has already put some work into defining this.

4.  **User-defined destructuring.** My favorite language features are ones that add new expressive syntax whose semantics can be programmed by an API designer. The `for-in` loop syntax is baked into the language, but you get to decide what it means in your implementation of Iterable.

    I want the same thing for destructuring. I think a syntax something like `Name(subpattern1, subpattern2)` (an identifier followed by a parenthesized list of subpatterns) should desugar to calling some user-defined destructuring operation on the matched value. That operation can define whether the value matches or not and, if it does, what subvalues the subpatterns are matched against. There are obviously many details to work out here, but also prior are in Scala's [`unapply()`][unapply] method and Kotlin's [`componentN()` methods][component].

[unapply]: https://docs.scala-lang.org/tour/extractor-objects.html
[component]: https://kotlinlang.org/docs/reference/multi-declarations.html

5.  **Data classes.** The least-defined (in my mind) goal is around improving the user experience for easily defining value-like types. This could mean some new explicit "data class" syntax, or a family of smaller features like allowing implicit constructors to take parameters.

    If we allow these two be subclasses of a superclass, then that gets pretty close to sum types once you include the previously-mentioned ability to pattern match on types and destructure them. Likewise, once you have general user-defined destructuring, then any lightweight data class-like syntax can just be syntactic sugar for a class declaration that includes a declaration of its own destructuring support.

As I said, this is all still very early and tentative, but I wanted to collect what I think are the related issues into one place and then begin sketching out a map through this territory.

**Comment by werediver** [comment]
A great digest 👍 In the meanwhile, there is a young library that generates sum-types (-alike classes) for you: https://github.com/werediver/sum_types.dart (is this shameless plug an off-topic?)

**Comment by rrousselGit** [comment]
If existing packages are interesting to see, here's a bunch of them:

- [sealed_unions](https://pub.dev/packages/sealed_unions)
- [tuple](https://pub.dev/packages/tuple)
- [dartz](https://pub.dev/packages/dartz)

**Comment by wkornewald** [comment]
Since you're collecting information, I'd like to mention [my "proposal"](https://github.com/dart-lang/sdk/issues/4938#issuecomment-396005754) where I suggested how you could combine pattern matching, data classes, union types, and typedef into a coherent whole while still making each of those concepts useful individually. The result is easier to use/understand **and** more flexible than algebraic data types. This is what we want.

So, pattern matching should IMHO focus on these orthogonal concepts of data classes (for destructuring) and union types (for exhaustiveness).

Also, before I found this issue I already [commented here](https://github.com/dart-lang/language/issues/83#issuecomment-531149172) on why algebraic data types result in a worse developer experience vs. union types. TL/DR: with algebraic data types, pattern matching forces you to check for cases at runtime (and raise exceptions) although they should be forbidden at compile-time! I suppose that's also relevant to this issue.

**Comment by werediver** [comment]
@wkornewald The untagged unions concept sounds nice, but it doesn't sound like a replacement to the algebraic data types. They can totally coexist and serve different needs.

(not clear which topic is more relevant, so put in both #546 and #83)

**Comment by wkornewald** [comment]
I'd say it's more relevant to #83, so let's discuss [over there](https://github.com/dart-lang/language/issues/83#issuecomment-531217941).

**Comment by modulovalue** [comment]
I haven't seen lenses mentioned anywhere before when data classes are being discussed. I strongly believe that they would greatly benefit from 'boilerplate free' lenses.
This package attempts to solve that by generating data classes and a lens for each property [functional_data](https://pub.dev/packages/functional_data). 
Another implementation for more advanced lenses can be found in the package dartz and an example can be found [here](https://github.com/spebbe/dartz/blob/master/example/lenses/example.dart).

**Comment by wkornewald** [comment]
@modulovalue That's a great suggestion. Lenses would be very helpful for safe and ergonomic state management (e.g. for my [flutter_simple_state](https://github.com/ensody/flutter_simple_state) package or, if you prefer, something similar to immutable.js + Omniscient in TypeScript). Unfortunately, it's seemingly impossible to define a lens that can wrap arbitrary (including 3rd-party) data structures / objects. Well, at least without some compile-time code generation...

Maybe lenses should be a separate issue, though?

**Comment by G5Becks** [comment]
@modulovalue Whats the benefit of lenses over record copying semantics of F# over lenses.
https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/copy-and-update-record-expressions ?

**Comment by spkersten** [comment]
I’m not familiar with it F#, but what is shown on the linked page looks
very similar to `copyWith` method. Lenses are much more powerful. They
allow you to change values deep within a nested data structure. See the
example in the functional_data package.

**Comment by modulovalue** [comment]
I'm also not too familiar with F# but as spkersten said, they are much more flexible by being a more general abstraction.

Take a look at the following code snippet:

```dart
class Foo {
  final Bar bar;
}

class Bar {
  final Baz baz;
}

class Baz {
  final String content;
}
```

A Lens of type `Lens<Foo, String>` removes the fact that a `Bar` exists. It lets me access the `content` of Foo by just having an instance of Foo. The consumer doesn't even care that Baz exists, or where content is. Any changes to the API would not break updateFoo.

```dart
Foo updateFoo(Lens<Foo, String> lens, Foo foo) {
  return lens.update(foo, "New Content");
}
```

Lenses also pave the way for many other abstractions like the `FocusedLens` in spkerstens package. Lenses can also be chained which I believe is not possible with copy and update expressions in F#.

But as I mentioned, creating such lenses is too verbose right now. I'd ideally want something like `Foo$.bar.baz.content`.

--
@wkornewald said
> Unfortunately, it's seemingly impossible to define a lens that can wrap arbitrary (including 3rd-party) > data structures / objects. Well, at least without some compile-time code generation...
> Maybe lenses should be a separate issue, though?

I don't see how the dart team could add support for lenses right now as doing that in the OOP context of Dart only makes sense for data classes. There's not a lot I can say other than asking politely to consider the strengths of this abstraction when implementing data classes.

**Comment by G5Becks** [comment]
@spkersten Very cool package, I didn't even know that existed! 

 @modulovalue 

I understand what lenses are, but I don't understand what the use case is at a language level vs baked in copyWith ? Even the most functional of languages don't have lens support at the language level. Haskell uses the lens library, Scala uses monocle, F# uses aether, Ocaml uses lens.

**Comment by jifalops** [comment]
Can someone comment on the layers that have an effect on the Dart language as described in the grammar to the Dart language I experience as a developer?

Another question: Would patterns allow default type parameters like `A<B=X, C=Y>`? Is this a thing? I want it to be a thing.

**Comment by burakemir** [comment]
I very much like the proposal to add a form of (ML style) algebraic data types - or even features that would let one encode algebraic data types in a convenient way.

I'd like to point out some "hooks" in the language spec that would make it easy to add ML style algebraic data types.

* The current spec (version 2.2) talks about enums (Section 13 Enums) being translated to classes; this is already very close to what a "data class" translation would be like. 

The following example uses a syntax that is borrowed from Scala 3 (see http://dotty.epfl.ch/docs/reference/enums/adts.html ...) but swift (https://docs.swift.org/swift-book/LanguageGuide/Enumerations.html) and Kotlin (https://kotlinlang.org/docs/reference/enum-classes.html) take a similar route.
```
enum Color {
  case Red
  case Green
  case Blue
  case Mix(int mix)
}

// alternatively:
enum Color(int rgb) {
  case Red extends Color(0xFF0000)
  case Green extends Color(0x00FF00)
  case Blue extends Color(0x0000FF)
  case Mix(int mix) extends Color(mix)
}
```

An enum either defines each case using the "case" keyword, or none (which would mean, it's a traditional enum as before). We'd extend the enum translation to create one subclass per case, e.g. (only doing the second example)
```
class Color {
  final int rgb;
  Color(this.rgb);
}

class Red extends Color {
   Red() : super(0xFF0000);
   // appropriate equals and hashcode implementations
}

class Mix extends Color {
  final int mix;
  Mix(this.mix) : super(mix);
  // ...
}
```

Instead of the verbose inline "extends", one could use the syntax for super constructors and say `case Red : super(0xFF0000)` right away. However it would be harder to specialize generic arguments, which is useful:

```
enum Expr<T> {
  case Number(int n) extends Expr<Int>;
  case String(string s) extends Expr<String>;
}
```
* Now, if we wanted data classes, it could make sense to introduce the "enum class"
```
enum class JustData(int foo, string bar) : SomeSuper(foo) {
  // normal class body, but ... no mutable fields
  JustData.extra(string fooArg) : this(_parse(fooArg))
}
```
Again, the same class translation that is spec'ed for "enum" would be used (which somewhat justifies the use of the num keyword).
```
class JustData extends SomeSuper {
  final int foo;
  final string bar; 

  JustData(this.foo, this.bar) : super(foo);
  JustData.extra(string fooArg) : this(_parse(fooArg))
  // appropriate equals and hashcode
  // normal class body
}
```

* A suitable syntax for patterns would include destructuring of case classes (whether they are in an enum or outside). Exhaustivity checking only applies to things that are defined within an enum (which takes care of "sealed" scenario). A pattern Foo(p1, ..., pN) would always match a Foo whose main constructor was called with e1, ..., eN and each e_i matches the p_i.

* a pattern matching expression (whether it's switch or match) should definitely be an expression, i.e. produce a value.

* a pattern matching expression can be translated in a nested tree of type tests and binding statements. We can add various forms of runtime tests to the pattern syntax (including type tests and and type binding, since that is available at runtime in dart). One can do an "optimizing" translation where the result of a type test is used for the rest of the match so never has to be needlessly repeated (it is possible to specify this as a requirement, however, it would be easier to leave it to implementations).

* Tuples are nothing but standalone enum classes with the appropriate number of components. (e1, ..., eN) gets desugared into TupleN(e1, ..., eN)
```
enum class Tuple2<T1, T2>(T1 part1, T2 part2);
```

 * About user-defined destructuring: a long time ago, when unapply was added to Scala, it was done via user-defined functions that would return Option<T> or Option<TupleN<T1, ..., TN>> for N >= 2 ... it would be good if such user-defined destructuring functions would not contain any side-effects, though.

```
enum Option<X> {
  case Some(X x)
  case None
}

enum Tree {
  case Leaf;
  case Node(Tree left, Tree right);
}

@unapply Option[Tree] isSpecial(Node tree) {
  return  tree.left == tree.right
    ? Some(tree.left)
    : None
}

exp match {
  case Leaf => ...
  case .isSpecial(tree) => ...  // we already know that "tree is Node"
  case Node(left, right) => ...
}
```

   * since the cases inside enum and also the "enum classes" are just classes, nothing special needs to be done for runtime representation.

I will leave it at that for now - hope this is helpful. Let me close with the statement that enum corresponds closely to the sum type from type theory (or *disjoint* union from set theory), so in my mind it's not a coincidence that all these languages have extended "enum" to cover case distinction in data modeling.

**Comment by munificent** [comment]
Thanks for the comment. I'm not sure if we literally want to build on Dart's existing enum syntax (though I'm not fundamentally opposed either). But, otherwise, we seem to be thinking in the same direction.

**Comment by g5becks** [comment]
Thought I’d share something I came across today while browsing through facebooks’ [skiplang](http://skiplang.com/) docs. [This](http://skiplang.com/docs/classes.html#children) might be an interesting approach for adt’s. Of course , using the keyword “children” wouldn’t be an option choice because of flutter and what not, but maybe something else would work. If nothing else, I think it’s worth a look.

**Comment by munificent** [comment]
Very interesting, thanks @g5becks!

**Comment by rrousselGit** [comment]
For some reason, extension methods helped quite a bit in that area. 

For example, we can write:
```dart
extension UnionValue<A> on Union2<A, A> {
  A get value => _value as A;
}

class Union2<A, B> {
  Object _value;
}
```

Which has interesting properties:


```dart
Union2<String, int> example;
// String and int have no shared interface, so value is of type Object
Object value = example.value;
int impossible = example.value; // COMPILE ERROR, not type safe

Union2<int, double> example2;
// int and double have a common interface: num. Therefore value is of type num
num value2 = example2.value; // valid and type safe
int stillImpossible = example2.value; // COMPILE ERROR, value is a num not a int

Union2<String, String> example3;
// The type is always a String, so value is a String too
String value3 = example3.value; // still type safe
```

See [union](https://github.com/rrousselGit/union) for a package using that pattern.



The only thing we're lacking for now is:
- being able to assign `Union2<String, int>` to `Union2<int, String>`
- being able to assign `Union2<String, int>` to `Union3<String, int, whatever>`

The first one is not too important. The most important one being the second one (Union2 to Union2+)

Maybe a simplification of the feature could be an "any"-like, so that we can do:

```dart
class Union2<A, B> implements Union3<A, B, any> {}
```

which would allow:

```dart
Union2<String, int> union2;
Union3<String, int, double> union3 = union2;
```

**Comment by g5becks** [comment]
@rrousselGit not sure how that fills the need for adt’s. E.G. how can you create a Result<T>  type without all the boilerplate it would take wrt the class hierarchy? I think the extension methods help fill the need of pattern matching for the interim, but not with the creation of what you are matching against. Or at least not for the use cases I think they would be used for mostly. 

From the looks of it (I could be wrong) a Union<A,A> that has multiple constructors is an And type and not an Or type? Looks more like a Tuple type than a Choice type.

**Comment by rrousselGit** [comment]
@g5becks No, this is not a Tuple but an actual union.
See [union](https://github.com/rrousselGit/union) 

TL;DR `Union2<String, int>` is manipulated with custom methods.
The typical switch case on the different types a union can take is performed this way:

```dart
Union2<String, int> myUnion;

myUnion.forEach(
  (String value) => print("got a string $value"), 
  (int value) => print("got an integer $value"), 
);
```


It also has an example of a result-like type https://github.com/rrousselGit/union#making-a-reusable-union

**Comment by g5becks** [comment]
@rrousselGit i should have read closer! That’s actually pretty cool. I think I’ll definitely be using this.

**Comment by rrousselGit** [comment]
Had another go at this recently.
Interestingly, we can write extensions for functions too.

Which means we can use functions instead of classes to implement our temporary unions.

Which means we can write:

```dart
Union2<String, int> union2;

Union3<String, int, double> union3 = union2; // upcast, no problem
Union4<String, int, double, Object> union4 = union2; // still works
```

while still being unable to write:

```dart
Union3<String, int, double> union3;

// downcast, does not work because the value can be a double too.
Union2<String, int> union2 = union3;
```


It's sad that Dart doesn't support well typedefs though. It'd be almost perfect: https://github.com/rrousselGit/union#custom-unions

**Comment by jogboms** [comment]
True, nice work! @rrousselGit 

Also patiently waiting on the `typedefs` of `typedefs` feature.

**Comment by xsahil03x** [comment]
I also faced these issues and handled using the factory constructors mentioned by @rrousselGit in a [StackOverflow post](https://stackoverflow.com/questions/54317000/are-there-any-sealed-classes-alternatives-in-dart-2-0) and was pretty satisfied with the results. To avoid the boilerplate, I built a [code generator](https://github.com/xsahil03x/super_enum) with my colleague which generates these classes by annotating Enums.

```dart
@superEnum
enum _MoviesResponse {
  @Data(fields: [DataField('movies', Movies)])
  Success,
  @object
  Unauthorized,
  @object
  NoNetwork,
  @Data(fields: [DataField('exception', Exception)])
  UnexpectedException
}
```
where:-

`@Data()` marks an enum value to be treated as a Data class.

- One should supply a list of possible fields inside the annotation.

- If you don't want to add fields, use `@object` annotation instead.

- Fields are supplied in the form of `DataField` objects.

- Each `DataField` must contain the `name` and the `type` of the field.

- If the field type needs to be generic use `Generic` type and annotate the enum value with `@generic` annotation.

`@object` marks an enum value to be treated as an object.

Then it can be used easily with the generated `when` function

```dart
 moviesResponse.when(
    success: (data) => print('Total Movies: ${data.movies.totalPages}'),
    unauthorized: (_) => print('Invalid ApiKey'),
    noNetwork: (_) => print(
      'No Internet, Please check your internet connection',
    ),
    unexpectedException: (error) => print(error.exception),
  );
```

**Comment by werediver** [comment]
@xsahil03x You may like another solution [mentioned earlier](https://github.com/dart-lang/language/issues/546#issuecomment-525172871) in this thread: [sum_types](https://github.com/werediver/sum_types.dart) (also on pub.dev). It has proven to be very useful in practice and (by now) it has noticeably less limitations then the solution outlined in your message. Hope you'll find it suitable for you needs!

**Comment by xsahil03x** [comment]
@werediver I tried `sum_types` but I didn't like the boilerplate we have to write for generating the classes and the boilerplate looks redundant to me. Whereas in `super_enum` we are using enums to replicate the same and it feels easy and not redundant at all as we are using the enum itself to generate exhaustive `when` statement.

As I am coming from the native development, annotating enums feel much similar to me than creating bunch of classes as boilerplate.

**Comment by werediver** [comment]
@xsahil03x There are reasons behind the design of `sum_types`. The boilerplate buys it compatibility with other code generators (e.g. [functional_data](https://pub.dev/packages/functional_data)) and extensibility (user-defined members). But surely you have your reasons behind the design of `super_enum`.

A less related feature of `sum_types` is that it facilitates arbitrary serialization-deserialization of the generated sum-types (as in [`Nat.fromJson()`](https://github.com/werediver/sum_types.dart/blob/f796b2590ebf10eb34e1e07d3b39f18746ada26f/example/lib/src/nat.dart#L12) / [`Nat.toJson()`](https://github.com/werediver/sum_types.dart/blob/f796b2590ebf10eb34e1e07d3b39f18746ada26f/example/lib/src/nat.dart#L15)). That can mostly be implemented with your approach as well (which is similar the approach of `sum_types` v0.1.x).

**Comment by Jonas-Sander** [comment]
Just gonna put something I had written somewhere here anyway 😉

I didn't yet quite grasp the advantages of this:
```dart
double calculateArea(Shape shape) =>
  switch (shape) {
    case Square(length: l) => l * l;
    case Circle(radius: r) => math.pi * r * r;
  };
```

From my point of view something like the following is simpler to understand:
```dart
double calculateArea(Shape shape) =>
  switch (shape) {
    shape is Square => shape.l * shape.l;
    shape is Circle => shape.r * shape.r * math.pi *;
  };
```

Related:
If you would want to have the name reflect the object type something like this may also work:
```dart
double calculateArea(Shape shape) =>
  switch (shape) {
    shape is Square (square) => square.length * square.length;
    shape is Circle (circle) => math.pi * circle.radius * circle.radius;
  };
```

**Comment by duncanmak** [comment]
@Jonas-Sander I think the mixing of pattern matching and destructuring together is a relic of the design from ML, which is very successful and the same design has been copied into other languages like Scala and Rust, etc.

But your proposal is an intriguing design, if we change the design by a bit, here's an idea that's partly inspired by COND in Scheme:

```
double calculateArea(Shape shape) =>
  switch (shape) {
    shape is Square: (square) => square.length * square.length;
    shape is Circle: (circle) => math.pi * circle.radius * circle.radius;
  };
```  

If we allow a lambda to come after the pattern, it could be cool to allow any arbitrary lambda in that spot, instead of just a literal block. That could work pretty nicely as well.

**Comment by Jonas-Sander** [comment]
Something more general:

I personally haven't yet felt the need for record types.  
This may very well stem from the fact that I didn't use any language where records are used very often (altough there are tuples in typescript - I've never had to use them).  

What I instead felt a real need for were:
* Simple/Quick data classes 
* Union types / Algebraic datatypes (I'm no expert with the nomenclature)

**Data classes / tuples / records**

Data classes are mostly "solved" with this (and the [record](https://github.com/dart-lang/language/blob/master/working/0546-patterns/records-feature-specification.md)) proposal, as the named record is more or less the same.  
For me it's just that with these proposal is so much stuff that I never felt like I needed. 
I've didn't really felt the need for tuples and I'm also kinda scared of this:
```dart
switch (obj) {
  case [int a, int b]:
  case {"a": int a, "b": int b}:
    print(a + b); // OK.

  case [int a]:
  case (String a): // Error.
    break;
}
```

For example instead of:
```dart
var (lat, long) = geoCode('Aarhus');
print('Location lat:$lat, long:$long');
```
I'd probably write:
```dart
// Similar to Kotlin just as an example
data class Location(final double lat, final double long);

final loc = geoCode('Aarhus');
print('Location lat:${loc.lat}, long:${loc.long}');
```
or with named parameters:
```dart
data class Location({final double lat, final double long});
```

Yes the record is shorter. For this example I like the record more.    
For a real codebase on the otherhand I would use the latter probably 95% of the time because I like that such a value object / data class is one place where I could put documentation etc.    
The real issue with Dart I have up until now is that I wish there would be a class with "value class"-ish `==` and `hashCode` implementation and maybe a `copyWith` autogenerated method. Not that I definitely need tuples.

This not to say tuples/records are a bad idea. Like already mentioned I'm personally not very familiar with them and because of this I also won't see all the use cases right for them.   
Some use cases I did see and where I think I might personally use records or tuples:
- `Future.wait` 
- A (exception, stacktrace) tuple (I think there are some methods where this could help? Not sure)

**Algebraic datatypes / Union types**

What I would really really like to see in Dart on the other hand is something I use in typescript _aaalll_ the time - union types.
For example:  `double calculateArea(Square | Circle shape) {...}`.
(Which are union types not algebraic datatypes, right?)

From the proposal:  
> Algebraic datatypes
> You often have a family of related types and an operation that needs specific behavior for each type. In an object-oriented language, the natural way to model that is by implementing each operation as an instance method on its respective type:
> 
> ```dart
> abstract class Shape {
>   double calculateArea();
> }
> 
> class Square implements Shape {
>   final double length;
>   Square(this.length);
> 
>   double calculateArea() => length * length;
> }
> 
> class Circle implements Shape {
>   final double radius;
>   Circle(this.radius);
> 
>   double calculateArea() => math.pi * radius * radius;
> }
> ```
> Here, the calculateArea() operation is supported by all shapes by implementing the method in each class. This works well for operations that feel closely tied to the class, but it splits the behavior for the entire operation across many classes and requires you to be able to add new instance methods to those classes.
> ....
> ```dart
> double calculateArea(Shape shape) =>
>   switch (shape) {
>     case Square(length: l) => l * l;
>     case Circle(radius: r) => math.pi * r * r;
>   };
> ```

I don't want to subclass. 
What I really like about `double calculateArea(Square | Circle shape) {...}` is that I can see what the method takes on the spot. 
Together with a `switch` expression this would be great:
```dart
double calculateArea(Square | Circle shape) =>
  switch (shape) {
    shape is Square (square) => square.length * square.length;
    shape is Circle (circle) => math.pi * circle.radius * circle.radius;
  };
```

Also what about the cases where the classes are unrelated?   
For example from what I can remember there are several Dart methods which take `someMethod(dynamic exceptionOrError)`.  
From what I can remember both are distinct classes. Could be `someMethod(Exception | Error exceptionOrError)`.  
Can this be experessed with Algebraic datatypes? (honest question I don't know)

In the end: I don't know if this stuff even made sense, I probably confused many ideas and concepts ;)
Just wanted to express my feelings regarding the proposal (which seems to be very early stage of course).

**Comment by Jonas-Sander** [comment]
@duncanmak 

For me it's just that when I see this I think of constructors:
```dart
case Square(length: l) => l * l;
```

> I think the mixing of pattern matching and destructuring together is a relic of the design from ML, which is very successful and the same design has been copied into other languages like Scala and Rust, etc.

Something I'm also not familiar with;)  
What are the advantages of ahving the pattern matching and destructuring together? There will be some but with the short example that is given I would probably use the easier syntax instead of introducing a new one into the language. Do you have some examples?

**Comment by Jonas-Sander** [comment]
(regarding my wall of text)

As I think the stuff regarding data classes / records is a bit hazy:  
I think in the end I would rather not introduce the records like proposed (and all the complexity) into the language but would rather have something like a data class which is a less complex concept (for me), may also integrate more nicely with automatic (de)serialization and is a single point for documentation etc.   

This may be because I don't understand records enough so I would love to see some use cases.

**Comment by tatumizer** [comment]
```
double calculateArea(Shape shape) =>
  switch (shape) {
    case Square(length: l) => l * l;
    case Circle(radius: r) => math.pi * r * r;
  };
```
The part `(length:l)` is effectively a destructuring (in general, there can be N parameters in the list), but it doesn't comply with the syntax of destructuring, and there's no such thing as destructuring of an arbitrary object in the proposal. It feels very ad hoc to me - like an artificial implant.

**Comment by burakemir** [comment]
The advantages of destructuring over "typecase" only becomes apparent when you nest patterns. For example:

Expr simplify(Expr expr) =>
  switch (expr) {
    case Plus(x, Zero()) => x
    case Plus(Zero(), x) => x
    case Mul(Zero(), x) => Zero()
    case Mul(x, Zero()) => Zero()
    case Mul(One(), x) => x
    case Mul(x, One()) => x
    default => expr
  }

Certainly in the academic context of ML (Hope, Miranda etc), manipulation of symbolic logic expressions  was very common, but despite the arithmetic expressions coming up as examples all the time, dealing with such tagged unions/records appears a lot in compilers, interpreters (DSLs), or implementing the GoF Interpreter pattern for things like user actions etc. where nested patterns are useful.

**Comment by cindRoberta** [comment]
related to https://github.com/dart-lang/language/issues/2006

**Comment by bouraine** [comment]
I would like to provide with an real example where discriminated union would be very useful to reduce the boilerplate and improve readability:

```
abstract class SearchState {}

class SearchInitialState extends SearchState {}

class SearchLoadingState extends SearchState {}

class SearchErrorState extends SearchState {
  final String errorMessage;

  SearchErrorState(this.errorMessage);
}

class SearchResultState extends SearchState {
  final List<String> result;

  SearchResultState(this.result);
}
```

These types are used with bloc-library to describe the state of the app at every step of the search:

```
class SearchBloc extends Cubit<SearchState> {
  SearchCubit() : super(SearchInitialState());

  void search(String term) async {
    emit(SearchLoadingState()); // UI updates after every emit to display the new state
    try {
      final result = await repository.search(term);
      emit(SearchResultState(result));
    } catch (ex) {
      emit(SearchErrorState(ex.toString()));
    }
  }
}
```

Flutter UI:

 ```
 return BlocBuilder<SearchBloc, SearchState>(builder: (context, state) {
    if (state is SearchLoadingState) {
        return LoadingWidget();
    }
    if (state is SearchErrorState) {
        return ErrorWidget(message: state.errorMessage);
    }
    if (state is SearchResultState) {
      return ResultWidget(result: state.result)  
    }
   ...
}
```

This could be replaced by something like: 
```
enum SearchState  {
  case Initial()
  case Loading()
  case Error(String message)
  case result(List<String> result)
}
```
And the pattern matching could be something like this in the the Flutter UI:
```
return BlocBuilder<SearchBloc, SearchState>(builder: (context, state) {
   return switch state {
      case Initial: SizedBox()
      case Loading: LoadingWidget()
      case Error: ErrorWidget()
      case Result: ResultWidget()
   }
}
```

**Comment by munificent** [comment]
> This could be replaced by something like the code below:
> 
> ```
> enum SearchState  {
>   case Initial()
>   case Loading()
>   case Error(String message)
>   case Result(List<String> result)
> }
> ```
> 
> > The code is concise and very clear. 7 LOC instead of 17 with spaces in the first example.

The [pattern matching proposal](https://github.com/dart-lang/language/blob/master/working/0546-patterns/patterns-feature-specification.md) doesn't currently include any nicer syntax for declaring a family of classes like you have here. I've put some thought into it, but not a lot. I agree it's very nice to have.

> And the pattern matching could be something like the code below in the the Flutter UI:
> 
> ```
> return BlocBuilder<SearchBloc, SearchState>(builder: (context, state) {
>    return switch state {
>       case Initial() => SizedBox()
>       case Loading() => LoadingWidget()
>       case Error(message) => ErrorWidget(massage: message)
>       case Result(result) => ResultWidget(result: result)
>    }
> }
> ```

Yes, that's very close to the current proposal. The exact syntax in the proposal would be:

```dart
return BlocBuilder<SearchBloc, SearchState>(builder: (context, state) {
   return switch state {
      case Initial() => SizedBox();
      case Loading() => LoadingWidget();
      case Error(message: var message) => ErrorWidget(message: message);
      case Result(result: var result) => ResultWidget(result: result);
   }
}
```

**Comment by Levi-Lesches** [comment]
> > This could be replaced by something like the code below:
> > ```
> > enum SearchState  {
> >   case Initial()
> >   case Loading()
> >   case Error(String message)
> >   case Result(List<String> result)
> > }
> > ```
> 
> The [pattern matching proposal](https://github.com/dart-lang/language/blob/master/working/0546-patterns/patterns-feature-specification.md) doesn't currently include any nicer syntax for declaring a family of classes like you have here. I've put some thought into it, but not a lot. I agree it's very nice to have.

Hopefully, this can be made simpler with #314, data classes:

```dart
abstract class SearchState { }
class Initial extends SearchState { }
class Loading extends SearchState { }
data class Error(String message) extends SearchState;
data class Result(List<String> result) extends SearchState;
```

Or even, until then, a simple macro from #1482, metaprogramming:
```dart
abstract class SearchState { }
class Initial extends SearchState { }
class Loading extends SearchState { }
@data class Error extends SearchState { final String message; }
@data class Result extends SearchState { final List<String> result; }
```

**Comment by cedvdb** [comment]
> Yes, that's very close to the current proposal. The exact syntax in the proposal would be:
> 
> ```dart
> return BlocBuilder<SearchBloc, SearchState>(builder: (context, state) {
>    return switch state {
>       case Initial() => SizedBox();
>       case Loading() => LoadingWidget();
>       case Error(message: var message) => ErrorWidget(message: message);
>       case Result(result: var result) => ResultWidget(result: result);
>    }
> }
> ```

How does it aggregate the different subclasses of `SearchState` if those are in different files, or even different packages ? Does the user have to explicitely aggregate them like in a group of some sort ?

Are this proposal and records likely to be a thing in 2022 or will static meta programing delay it too much for that to happen ? If that's the case that really a bummer, I'd take this over static meta programing any day.

**Comment by munificent** [comment]
> How does it aggregate the different subclasses of `SearchState` if those are in different files, or even different packages ? Does the user have to explicitely aggregate them like in a group of some sort ? Or does the analyzer gets every subclass of SearchState it can reach ?

For non-exhaustive matches, there's no need to explicitly know or collect all of the subtypes. You can just check against whatever types you want. In order to have sound exhaustiveness checking, you need to know that the set of subtypes is a *closed* set. The proposal doesn't have any support for that yet. It's something I'm hoping to work on soon. So I guess a fully correct example using the proposal today would be more like:

```dart
return BlocBuilder<SearchBloc, SearchState>(builder: (context, state) {
   return switch state {
      case Initial() => SizedBox();
      case Loading() => LoadingWidget();
      case Error(message: var message) => ErrorWidget(message: message);
      case Result(result: var result) => ResultWidget(result: result);
      default: throw 'Unexpected type.';
   }
}
```

**Comment by rrousselGit** [comment]
> The proposal doesn't have any support for that yet.

Exhaustive switch case is probably the most important part of this issue I'd say

Otherwise, especially with Metaprogramming, I'd likely still end up using Freezed unions instead of the switch-is syntax, just for the fact that it's `when` is exhaustive

**Comment by bradyt** [comment]
> ... the set of subtypes is a _closed_ set.

Oops! I was stuck on this for a moment. I thought you meant https://en.wikipedia.org/wiki/Closed_set, but I'm guessing you mean like https://en.wikipedia.org/wiki/Closure_(mathematics).

I was wondering what are the open sets on the set of types, but I guess you just mean like a set closed under subtyping.

Okay I see now:

> A set that is closed under this operation is usually referred to as a closed set in the context of topology. Without any further qualification, the phrase usually means closed in this sense.

I was almost writing to ask if subtyping was like a limit point somehow.

**Comment by rrousselGit** [comment]
> I think having patterns and being able to match on them at all is probably the most important part.

We definitely are in an agreement :smile:

Rather, I think the switch-case proposal is optional all things considered.
Sure, having destructuring + expression switch-case + `case is Type` are all neat features. But in the grand scheme of things, they are only syntax sugar and don't add anything that we can't already do with:

```dart
if (state is Initial) {

} else if (state is Loading) {
} ...
```


My point being, say we were to split the various features that this issue involves across different Dart versions, I'd prefer having the ability to define sealed-class equivalent first.

My reasoning is, even without destructuring and all the other fancy features proposed here, if we had something similar to sealed-classes, we could make our `if (state if A) else if (sate is B)` an exhaustive check already.
Dart already does that with null-safety after all.

Such that we'd have:

```dart
// ignore the syntax, not what's discussed here
sealed class Interface {
  class A {}
  class B {}
}


int function() {
  Interface value = <whatever>;
  
  if (value is A) {
    return 0;
  } else if (value is B) {
    return 1;
  }
  // Never reached, and the compiler knows it.
  // As such, Dart isn't complaining that the function is lacking a return
}
```


This way we'd have exhaustive pattern matching already.  
The other parts of the proposal could wait IMO as long as we can do that.

**Comment by munificent** [comment]
Sealed class hierarchies is also quite a bit of work. Anything that touches modifiers and restrictions on how classes can be used opens up a can of worms around whether API consumers or maintainers should have more control.

I could be wrong, but I think exhaustiveness checking is a relatively minor feature. It's important if you want to rewrite a *lot* of your code into an ADT style. But for most typical Dart code which will likely remain fairly object oriented, I think it's in the same general area of importance as unreachable code warnings. Useful to prevent bugs, but not life-changing.

**Comment by twop** [comment]
> I could be wrong, but I think exhaustiveness checking is a relatively minor feature. It's important if you want to rewrite a lot of your code into an ADT style. But for most typical Dart code which will likely remain fairly object oriented, I think it's in the same general area of importance as unreachable code warnings. Useful to prevent bugs, but not life-changing.

Personal opinion: I think exhaustiveness for ADTs is almost inseparable part of the feature. I think the biggest appeal to use ADTs is to be able to model your logic more accurately AND outsource verifying the correctness of it to the compiler. 

That unlocks this workflow: I add/remove variants and rely on the compiler to guide me where I need to change the code.

from my perspective, removing the compiler support for exhaustiveness makes that mental "outsourcing" either impossible or untrustworthy (which is probably even worse?)

Side note:
I wonder what memory models ADTs in Dart can support? E.g. can ADT's values be allocated on the stack or heap only? F# has support for stack allocated ADTs but using them for high perf compute present a big challenge for CPU cache lines (they are too big), Rust on the other hand allocates them on the stack with the size of the "biggest" variant which is way more efficient.

**Comment by rrousselGit** [comment]
> Useful to prevent bugs, but not life-changing.

I'd argue that them preventing bugs is the very reason why they are life changing.

Isn't the whole point of using pattern matching to prevent bugs. If it doesn't prevent bugs, why bother? 

That's why Freezed offers an exhaustive check.  
That's also why some folks made PRs to Freezed to disable the generation of the non-exhaustive pattern matches
https://github.com/rrousselGit/freezed/issues/562


That issue was essentially them saying "My team prefers having a worse syntax for the sake of exhaustive checks so that we can prevent bugs", which I totally get behind.

**Comment by mraleph** [comment]
> but I think exhaustiveness checking is a relatively minor feature.

I have said this internally, but I want to also say this externally as well. 

I think we should not be considering features like pattern matching, data classes, sealed class hierarchies etc in separation.  Rather they should all be viewed as a part of a single package. I like to call this package _features enabling functional domain modeling_, because it reflects that use case that it is going to address. 

You get most of the benefits from these features when you combine all of them in a single cocktail where you get static checking, IDE completion (e.g. you write `switch state` and strictly speaking IDE should over you to create an exhaustive switch), ergonomics, etc. 

In this picture of the world exhaustiveness is not a minor subfeature, but rather a crucial piece of an overall offering without which the overall offering looses a lot of the benefits. 

I could also draw a parallel between exhaustiveness checking and NNBD, because they bring similar guarantees around safety and making code simpler to reason about.

**Comment by bouraine** [comment]
I am of the same view as [this comment of @eernstg ](https://github.com/dart-lang/language/issues/83#issuecomment-449380965)
enum mechanism with associated values would be a quite natural way in Dart
I 'd prefer enums to sealed classes

**Comment by cedvdb** [comment]
I've been wondering why this kind of pattern is more prevalent on the client side. On one end it's more correct and more provable from a compilation perspective, in a sense, on par with null safety. On the other, making all these state classes requires extra care, more boiler plate and a bit of a new way to think about OOP.

Moreover this is especially prevalent in the UI layer than any other layers of a client application. 

It's basically dynamic polymorphism. This class has the data property when `load` method resolved successfully. The provability under compilation is what I find interesting.  

The provability part is what I find really interesting and have been wondering if there was a way that could be extended to other layers without making the code too obtuse. Maybe some hints to the compiler could help the provability part without requiring too much boiler plate.

```dart
class DataLoader {
   @not-null-when(isLoadingSuccessful == true)
   Data? data;
   bool isLoadingSuccessful
}

// requiring
if (dataLoader.isLoadingSuccessful)
  useData(dataLoader.data);
}
```

**Comment by rrousselGit** [comment]
> I am of the same view as this comment of eernstg : enum mechanism with associated values would be a quite natural way in Dart. I'd personally prefer enums over sealed classes.


If enums can be non-constant, I don't think they can be considered enums anymore and are rather normal classes.

**Comment by mraleph** [comment]
@cedvdb the boilerplate is the artefact of current syntactic decisions. `enum classes` reduce amount of boilerplate you need to write and allow capturing the state invariants easily. 

```
enum DataLoaderState {
  loading, 
  loaded(Data data),
}
```

> If enums can be non-constant, I don't think they can be considered enums anymore and are rather normal classes.

One can be viewed as a natural extension of another though. Enum cases without any arguments (or with constant arguments) can be represented by constant objects.

Unfortunately I think enhanced enums are taking the language on a different path which means it would be harder to consolidate this in the future.

**Comment by bouraine** [comment]
> If enums can be non-constant

Why do you thing that they should be non-constant ?

**Comment by rrousselGit** [comment]
> Why do you thing that they should be non-constant ?

They'd have to if we want to store the result of a network request in an enum if we want to use enums as an enabler for pattern-matching.

> One can be viewed as a natural extension of another though. Enum cases without any arguments (or with constant arguments) can be represented by constant objects.

On paper, yes. But I could see that cause some confusion, with people asking "Why use a class when I can use an enum" or those sorts of questions.

Especially with Enhanced Enums in mind, it'd feel like we'd have two keywords to do the same thing (enum vs class), but with enums having a better syntax.

After-all, lots of people are asking for data-classes. Yet the answer to this request seems to be "We won't be doing that as part of the language, but rather through code-generation".  With the current talks being about having the constructors generated – which is a syntax degradation.

Yet at the same time we're talking about allowing enums to do:

```dart
enum AsyncValue<T> {
  loading,
  Error(Object? error, [StackTrace? stackTrace]),
  Data(T value),
}
```

which is essentially data classes + sealed classes. 


That feels paradoxical to me. If that syntax is reasonable, then I don't understand why true data classes would be undesirable:

```dart
data class Model<T>(int count, {T? value});
```

**Comment by Nikitae57** [comment]
Please, we need this feature

**Comment by munificent** [comment]
> I'll insert that I think it would be cool if the team could go through a bunch of the "smaller" features, like data classes (#314), automatic constructors (#698), number delimiters (#1),

The "size" of a feature can be non-obvious sometimes. Syntactic features like data classes and enhanced default constructors are relatively small in terms of implementation effort (which is great), but they can be surprisingly large in terms of *design* effort. Figuring out *what* the best syntax should be is really hard and once we pick we're stuck with it so we want to make sure we get a design that's great.

**Comment by munificent** [comment]
For those following along, I wrote a [long doc](https://github.com/dart-lang/language/blob/master/working/0546-patterns/why-two-kinds-of-patterns.md) explaining the approach the current proposal takes for distinguishing refutable and irrefutable patterns.

**Comment by OlegAlexander** [comment]
@munificent Wow, what a great document. Thank you for putting so much thought into this feature! I agree with your rationale that Swift-like pattern matching is appropriate for Dart. If `switch` remains a statement, are you considering adding a `match` expression?

**Comment by munificent** [comment]
> If `switch` remains a statement, are you considering adding a `match` expression?

[The proposal](https://github.com/dart-lang/language/blob/master/working/0546-patterns/patterns-feature-specification.md) builds on the existing switch to allow patterns in switch statements but it also separately adds support for [switch expressions](https://github.com/dart-lang/language/blob/master/working/0546-patterns/patterns-feature-specification.md#switch-expression) too.

**Comment by munificent** [comment]
More documentation! If you've found the proposal a little too large to wrap your head around, I've added [a little summary](https://github.com/dart-lang/language/blob/master/working/0546-patterns/patterns-feature-specification.md#grammar-summary) of all of the kinds of patterns and where they are allowed to appear.

**Comment by cedvdb** [comment]
Thank you for the summary!

Has there been discussion about whether the `case` keyword should stay ? It seems superfluous

**Comment by munificent** [comment]
> Has there been discussion about whether the `case` keyword should stay ? It seems superfluous

Yes, some. I lean towards keeping most of the current syntax including `case` because it's familiar. It's hard to eliminate that without potentially running into ambiguity because each case contains an undelimited list of statements. That means whatever syntax we use to separate cases needs to not overlap either statement syntax or (thanks to expression statements) expression syntax. That's a pretty crowded grammar.

It's not impossible to eliminate `case` or find something shorter, but it's not trivial either.

**Comment by Jonas-Sander** [comment]
While I got this pretty quickly

```dart
var list = [1, 2, 3];
var [a, b, c] = list;
print(a + b + c); // 6.
```

on the other hand this

```dart
map = {'first': 1, 'second': 2};
var {'first': a, 'second': b} = map;
print(a + b); // 3.
```
threw me off at first.   
Especially `var {'first': a, 'second': b}` was confusing for me to understand because:
* usually the variable names are not wrapped in so much code (hard to scan over the code quickly)
* the normal way of using variables is turned around (usually `{first: a}` would mean that I assign the value of `a` to `first` but now it's the other way around).   

### switch

```dart
double calculateArea(Shape shape) =>
  switch (shape) {
    case Square(length: l) => l * l;
    case Circle(radius: r) => math.pi * r * r;
  };
```

I think I wrote this before but I don't like that this `Square(length: l)` looks like a constructor.   

It also works in the exact opposite way of normal constructors where `l` is a value you provide but in this case `l` is a variable that is assigned to by the statement to use after the `=>`. I think this would make the language harder to understand.

I want the switch statement to switch on class types in a concise and exhaustive way. I don't need a shorter way to access the attributes (at least not in the proposed syntax). It's fine for me to write more chars but avoid this confusing syntax.

```dart
double calculateArea(Shape shape) =>
  switch (shape) {
    // 1. just use the var in the switch
    case Square => shape.length * shape.length;
    // 2. use a closure to name the object to make it more readable (could be optional)
    case Circle (circle) => math.pi * circle.radius * circle.radius;
    // 2.1 closure can be used to save at least some code
    case Square (s) => s.length * s.length;
    // 3. Use other kind of brackets to signify that it's not a constructor
    case Circle[radius: r] => math.pi * r * r;
    // 4. Make the assignment clearer?
    case Circle[r = radius] => math.pi * r * r;
    // 4.1 Or maybe like this?
    case Foo(b = bar; q = quz) => bar - quz;
  };
```

Also in the proposal was
```dart
display(Object obj) {
  switch (obj) {
    case Rect(var width, var height):
      print('Rect $width x $height');
```
I think the `var` differentiates it a bit more from a constructor but I still don't like it that much 😁. 


### Implicit break

Hell yeah, lets get rid of these `break` statements!

### Switch expression

I love it!  
I also like making it high precedence.    

### switch-case / if-case statement

> Often you want to conditionally match and destructure some data, but you only want to test a value against a single > pattern. You can use a switch statement for that, but it's pretty verbose:
> 
> ```dart
> switch (json) {
>   case [int x, int y]:
>     return Point(x, y);
> }
> ```

Would `case [int x, int y]:` only match if `json` is a list with two int values? What about three ints? Or does it work in a whole other way and I don't get this whole proposal at all yet?:D   

Just for my undestanding: The `:` could also be replaced with `=>` in this example right?
```dart
// returns Point or null
return switch (json) {
  case [int x, int y] => Point(x,y)
}
```
> We can make simple uses like this a little cleaner by introducing an if-like form similar to if-case in Swift:
> ```dart
> if (case [int x, int y] = json) return Point(x, y);
> ``` 

Thats really confusing for me!    
The way I reason about `if`, `case` and `=` from normal Dart code just make no sense here anymore.  
If there is any `if` then I expect an `==` but this is assignment and comparison at once. Also adding `case` just adds to the confusion for me. I don't know Swift, I just know Dart and this confuses me.

### typeOrBinder rule - final

> The `typeOrBinder` rule is similar to the existing type grammar rule, but also allows final followed by an identifier to declare a type variable.
> ```dart
> switch (object) {
>   case List<final E>: ...
>   case Map<String, final V>: ...
>   case Set<List<(final A, b: final B)>>: ...
> }
> ```

Can someone explain what `final` is used for here in simple terms? I didn't understand that unfortunately.

### destructuring named fields

> When destructuring named fields, it's common to want to bind the resulting value to a variable with the same name. As a
> convenience, the binder can be omitted on a named field. In that case, the field implicitly contains a variable binder
> subpattern with the same name. These are equivalent:
>
> ```dart
>   var (first: first, second: second) = (first: 1, second: 2);
>   var (first:, second:) = (first: 1, second: 2);
> ```

Why not just:
```dart
var (first, second) = (first: 1, second: 2);
var (bar, foo) = (foo: 1, bar: 2);
```

I guess there is probably some ambiguity?

-------

I like that this proposal exists and that we think about new ways to make the language better and exiting.   
On the other hand I don't know it just seems like such a mountain of complexity and just hard to get into (at least for me).

**Comment by munificent** [comment]
> I think I wrote this before but I don't like that this `Square(length: l)` looks like a constructor.
> 
> It also works in the exact opposite way of normal constructors where `l` is a value you provide but in this case `l` is a variable that is assigned to by the statement to use after the `=>`. I think this would make the language harder to understand.

This is basically how patterns work in every language that has them. The pattern syntax mirrors the corresponding expression syntax that would create an object that the pattern destructures. It's weird at first, but once you get used to the idea, it lets you easily infer what a pattern means based on what you know about the corresponding expression.


> Would `case [int x, int y]:` only match if `json` is a list with two int values? What about three ints?

Just two int values. The proposal for list patterns requires it to have exactly the matched length. (We'll probably add a `...` syntax to mean "match a list of any length that starts with these elements".)

> Can someone explain what `final` is used for here in simple terms? I didn't understand that unfortunately.

It lets you extract the runtime type argument out of an instance of a generic type:

```dart
// The static type is List<Object>, the actual runtime type is List<int>:
List<Object> list = <int>[];

switch (list) {
  case List<final T>(): print('Runtime type is $T'); // Prints "int".
}
```

There's currently no way in the language to extract a generic object's reified type arguments.

> Why not just:
> 
> ```dart
> var (first, second) = (first: 1, second: 2);
> var (bar, foo) = (foo: 1, bar: 2);
> ```
> 
> I guess there is probably some ambiguity?

Yes, since a record can contain both named and positional fields.

**Comment by lrhn** [comment]
> This is basically how patterns work in every language that has them. The pattern syntax mirrors the corresponding expression syntax that would create an object that the pattern destructures. It's weird at first, but once you get used to the idea, it lets you easily infer what a pattern means based on what you know about the corresponding expression.

Letting construction and destructuring mirror each other's syntax is also the reasoning behind the structure of [C types](http://c-faq.com/decl/spiral.anderson.html). I'm not sure that's necessarily a convincing argument for the result being readable :wink:.

The pattern syntax here does work fairly well for predictably structured values with good literal syntax, like lists and types themselves.
It can work for maps, if you provide the keys in the pattern, but it can easily get hairy because `{`/`}` means so many things already. 
Will probably never be able to match sets, because they have no predictable structure and no way to enforce an ordering in a predictable way (or at least, anything which works for sets could also work for arbitrary iterables).

Dart objects are not Haskell/SML abstract data-types, they have properties separate from the constructor arguments/field values, and that abstraction has to be respected. You can't always mirror construction there.

**Comment by munificent** [comment]
> Letting construction and destructuring mirror each other's syntax is also the reasoning behind the structure of [C types](http://c-faq.com/decl/spiral.anderson.html). I'm not sure that's necessarily a convincing argument for the result being readable 😉.

Oh you don't have to tell me that "declaration reflects use" was a misfeature. :)

> The pattern syntax here does work fairly well for predictably structured values with good literal syntax, like lists and types themselves. It can work for maps, if you provide the keys in the pattern, but it can easily get hairy because `{`/`}` means so many things already.

Yeah, one of the challenges I ran into was deciding how named record field destructuring works. Most ML-derived languages treat tuples and records as totally distinct concepts with different syntax for each. They use `{ ... }` for records and then use that for record patterns too. But that would be weird in Dart because we use `{ ... }` for maps. Also, my proposal for records unifies positional and named fields into a single construct. That's why I went with `(...)` for destructuring even named fields.

> Will probably never be able to match sets, because they have no predictable structure and no way to enforce an ordering in a predictable way (or at least, anything which works for sets could also work for arbitrary iterables).

Yeah, I thought about sets and didn't come up with anything. In practice, I think it's fine. I don't think they're used that often and destructuring them feels like an odd operation since they're unordered.

> Dart objects are not Haskell/SML abstract data-types, they have properties separate from the constructor arguments/field values, and that abstraction has to be respected. You can't always mirror construction there.

Agreed that there's no perfect correspondence and no perfect syntax. We just have to try to come up with the best we can building on the grammar we already have.

**Comment by cedvdb** [comment]
I can get behind that it will become second nature but I wouldn't mind a keyword either. 
 
`case Square with (length: l) => l * l;`

Maybe it's cleaner without keyword once you get used to it though

**Comment by Levi-Lesches** [comment]
> That's why I went with `(...)` for destructuring even named fields.

I also like this because the proposal says to think of records like a reified version of a function's parameter list, so making the syntax match as closely as possible helps with that.

**Comment by munificent** [comment]
> I guess because devs can import and implement any classes that aren't private?

It's this. The supertype needs to be explicitly marked sealed so that it can be an error for downstream code to implement or extend the class.

**Comment by SandroMaglione** [comment]
How is this issue related to Static Metaprogramming #1482 ?
Is the dart team planning to work on both of this features? Are they related somehow? Is it possible that Static Metaprogramming could overlap in some of its capabilities to patterns?

Also, is there a timeline estimation for patterns to be available in dart ([similar to this estimation for Static Metaprogramming](https://github.com/dart-lang/language/issues/1482#issuecomment-1148936895))?

**Comment by munificent** [comment]
> How is this issue related to Static Metaprogramming #1482 ?

It's not. :)

> Is the dart team planning to work on both of this features?

Yes! We have a big enough language team to do some work in parallel.

> Are they related somehow? Is it possible that Static Metaprogramming could overlap in some of its capabilities to patterns?

In general, yes, we always hope that every new language feature can compose with other features in useful ways. But these two features specifically don't have a lot of affinity for each other as far as I know. Users may certainly surprise us with what they come up with, though.

**Comment by ykmnkmi** [comment]
Can patterns be used in function definitions?
```dart
void printPoint(Point(x: x, y: y)) {
  print('x: $x, y: $y');
}

**Comment by munificent** [comment]
The current proposal does not allow patterns in parameter lists, unfortunately. It's something I would really really like to do but... Dart's parameter list syntax is already *super* weird and complex. You have `required`, `this.`, `super.`, `[...]` for optional parameters, `{...}` for named parameters, default values, and probably other stuff I'm forgetting. Figuring out a pattern syntax that also harmonizes with all of that is very hard.

I expect that our initial release of pattern matching won't support patterns in parameter lists but I do hope we can get there eventually once we have time to work out all the details.

**Comment by Levi-Lesches** [comment]
Follow-up question: Could you be able to unpack a record to use as arguments to a function? I didn't see anything in the [records proposal](https://github.com/dart-lang/language/blob/master/working/0546-patterns/records-feature-specification.md) describing it, but it does say it "is identical to the grammar for a function call argument list."

```dart
void printPoint(int x, int y) {
  print('x: $x, y: $y');
}

void main() {
  final (int, int) origin = (0, 0);
  printPoint(origin);  // some special syntax to unpack the record?
}
```

**Comment by munificent** [comment]
> Follow-up question: Could you be able to unpack a record to use as arguments to a function? I didn't see anything in the [records proposal](https://github.com/dart-lang/language/blob/master/working/0546-patterns/records-feature-specification.md) describing it, but it does say it "is identical to the grammar for a function call argument list."

The current proposal does not have any support for spreading records into argument lists. That has always been a goal of mine and still is, but I suspect we won't get it into the first release of support for records and patterns. The proposal already has a *lot* of surface area, and that's something we can reasonably add later, I think.

**Comment by SandroMaglione** [comment]
Is the dart team planning to work on Patterns before, after, or at the same time of #1482? When would it be reasonable to expect a first alpha implementation of this?

**Comment by mateusfccp** [comment]
> Is the dart team planning to work on Patterns before, after, or at the same time of #1482? When would it be reasonable to expect a first alpha implementation of this?

They are working on both, as of now, although more recently I've seen much more movement in patters than static meta-programming.

I can't say about release, tho. Maybe they are going to release both on Dart 3?

**Comment by bouraine** [comment]
`records` has recently moved to [/accepted/feature-releaase]( https://github.com/dart-lang/language/tree/master/accepted/future-releases
)

**Comment by munificent** [comment]
> Is the dart team planning to work on Patterns before, after, or at the same time of #1482?

Records and patterns will be coming before static metaprogramming, unless something comes up that causes us to change the plan.

> When would it be reasonable to expect a first alpha implementation of this?

Sorry, but we don't generally offer future release dates since dates tend to change often.

**Comment by xmine64** [comment]
> Records and patterns will be coming before static metaprogramming

Even knowing that they'll come is so good. Without them codes are really dirty and and running build_runner takes so much time.

**Comment by atrauzzi** [comment]
I just wanted to add my 2c, but you guys really need to look at how C# has implemented pattern matching and destructuring.  It's turned out to be so elegant at least in utility.

```csharp
if (something is { YouCan: true, Do: { Some: "crazy", Stuff: true}})
{
   // ...
}
```
Even if it looks verbose at first blush, the amount of verbosity it cuts down when performing checks of complicated object hierarchies is just priceless.

**Comment by munificent** [comment]
> you guys really need to look at how C# has implemented pattern matching and destructuring.

We did. :)

> It's turned out to be so elegant at least in utility.
> 
> ```cs
> if (something is { YouCan: true, Do: { Some: "crazy", Stuff: true}})
> {
>    // ...
> }
> ```

Yes, the proposal supports arbitrary destructuring like that. There are some syntactic differences between Dart has map literals which naturally want to claim the curly braces, but the level of expressiveness is roughly the same.

**Comment by pedromassango** [comment]
@munificent now that this is being implemented how the new "Data class" feature will look like? Or perhaps where it is being dicused/designed?

**Comment by munificent** [comment]
No major updates on data classes. One relevant bit is that part of the design of views includes a primary constructor-like syntax and that led to discussing whether we can generalize that for classes (#2364). I believe we can and my hope is that we'll be able to design and ship that at some point after we ship records and patterns. My feeling is that primary constructors is a big piece of what users want when they ask for data classes.

**Comment by SandroMaglione** [comment]
@munificent since the whole "Pattern and related feature" is being implemented, does this include all the feature mentioned in this PR? I suppose that Record will be the first feature, but also the others are work in progress? Specifically, what about [ADTs](https://github.com/dart-lang/language/issues/349)?

**Comment by munificent** [comment]
I can't promise it covers every single comment in this issue (there's a lot), but, yes, we're doing records and [pattern matching](https://github.com/dart-lang/language/blob/master/accepted/future-releases/0546-patterns/feature-specification.md) and implementing both now. That includes support for [algebraic datatype-style code](https://github.com/dart-lang/language/blob/master/accepted/future-releases/0546-patterns/feature-specification.md#algebraic-datatypes).

**Comment by blueturningred** [comment]
I've played around with Dart 3 Alpha and it's such a major upgrade in DX. Great job everyone involved!

@munificent With sealed classes in places, is it there a plan in the future to implement narrowing with the ternary operator when working with sealed classes? Something like this:

```
sealed class User {}
class A extends User {
  final String firstName;

  A(this.firstName);
}

class B extends User {
  final String lastName;

  B(this.lastName);
}

String formatName(User x) {
  return x is A ? x.firstName : x.lastName // Dart 3 can't deduce that x must be of Class B at this point.
}
```

That would be really valuable for handling conditional widget rendering inline in Flutter without the need to create a function with a switch statement.

**Comment by lucavenir** [comment]
@blueturningred I may be wrong, but why not just implementing the `switch` operator instead? Dart will never know how many subclasses will inherit `sealed class User`, therefore using `switch` should be more safe and fully exhaustive.


Reading `return x is A ? x.firstName : x.lastName` isn't really readable either. What is `x`'s type in the "else" side of the ternary operator? Why should we able to acces `lastName` in an `else` statement, that should be the equivalent of a `default` case in the above `switch` operator? If anything, a `default` clause should let us access `User` fields, but not just any other subclass field.

Maybe I'm missing something here.

**Comment by lucavenir** [comment]
> You have two subclasses

Sorry if I'm bothering you, but what happens if you add a third one? Could you show that with Typescript, too?

**Comment by blueturningred** [comment]
> Sorry if I'm bothering you, but what happens if you add a third one? Could you show that with Typescript, too?

Not at all. If you add a third one which doesn't have `lastName`, then the compiler would complain and you'd be required to go back to formatName and account for the third one. The error in this particular case would be:

```
Property 'lastName' does not exist on type '{ _type: "UserB"; lastName: string; } | { _type: "UserC"; anothProp: string; }'.
  Property 'lastName' does not exist on type '{ _type: "UserC"; anothProp: string; }'
```
You can check the code [here](https://www.typescriptlang.org/play?#code/C4TwDgpgBAqgzhATlAvFA3lA+qSAuKAcniQEFCAaKAMwEtE5gA5AQwFsIDHFaA7AcwC+UAD4Yc4TkRKIAQpSgAbFo1YcuwHgOFj0E-NISIAwgpa8A9sAAWABUQWwGrUIBQr6gFdeAY2C0LXhoLRDYWZnYIAApPIwIZAEoMVyhUqEQIYE9EINikADp9aBQSwzJCKAB+KDzEfLoGCI4oAlr85VVI10EgA).

**Comment by lucavenir** [comment]
Thank you @blueturningred, I understand this now. It's even better to what `@freezed` does today.

**Comment by munificent** [comment]
> @munificent With sealed classes in place, is it there a plan in the future to implement narrowing with the ternary operator when working with sealed classes?

Good question! The main issue to discuss this is #2179. We aren't currently planning to have type promotion take sealed types into account and at this point it's very unlikely it will make it into 3.0. I think that's *probably* the right call. Flow analysis and type promotion versus sealed types and pattern matching are sort of two opposing styles for solving the same problem (moving from a more general type to a more specific one).

Most languages don't let you mix those styles *at all*. Rust, Swift, Haskell, and other functional languages do pattern matching but don't do any flow analysis. Kotlin and TypeScript do flow analysis but don't really have pattern matching. Dart is a little odd in doing both but it can get kind of weird when the two styles mix. I think it's probably easier for users to understand if we try to keep those styles mostly confined.

I could be wrong. But for now, if you want to work exhaustively with sealed types, that's what switches are for. If you want to work with type tests and control flow, that's what type promotion is for.

As to your example, I would write:

```dart
sealed class User {}

class A extends User {
  final String firstName;

  A(this.firstName);
}

class B extends User {
  final String lastName;

  B(this.lastName);
}

String formatName(User x) =>
  switch (x) { A a => a.firstName, B b => b.lastName };
```

**Comment by munificent** [comment]
Closing this because records and patterns are now enabled by default on the main branch of the Dart SDK! 🎉

**Comment by albertodev01** [comment]
@munificent you should move this into the "Done" column of the "Language Funnel" project

**Comment by munificent** [comment]
I'm not sure exactly how the timing for stuff in the language funnel is managed. That might wait until it ships on stable. I'll let @mit-mit or @itsjustkevin handle that part. :)

**Comment by talski** [comment]
will I be able to use the record identifier instead of indexes?

```dart
const (String name, int age) person = ('John doe', 32);

// this
print('${person.name}, ${person.age}');

// or this
print('${person.$name}, ${person.$age}');

// is more readable than this
print('${person.$1}, ${person.$2}');
```

**Comment by SandroMaglione** [comment]
@talski Yes, you can do as follows:

```dart
const (String name, int age) person1 = ('John doe', 32);
print('${person1.$1}, ${person1.$2}');

const ({String name, int age}) person2 = (name: 'John doe', age: 32);
print('${person2.name}, ${person2.age}');
```

You can read more on [Record type annotations](https://github.com/dart-lang/language/blob/master/accepted/future-releases/records/records-feature-specification.md#record-type-annotations):

> ```dart
> (int, String name, bool) triple;
> ```
> Each field is a type annotation and an optional name which isn't meaningful but is useful for documentation purposes.
>
> Named fields go inside a brace-delimited section of type and name pairs:
> ```dart
> ({int n, String s}) pair;
> ```

**Comment by purplenoodlesoop** [comment]
This feature is getting more exciting version by version! 

Will this type of inference be possible in further iterations? Currently, the following code produces a `A value of type 'int' can't be returned from the function 'eval' because it has a return type of 'T'` error:

```dart
sealed class Sealed<T> {}

class Case implements Sealed<int> {}

T consume<T>(Sealed<T> value) => switch (value) {
      Case() => 1,
    };
```

**Comment by lucavenir** [comment]
@purplenoodlesoop To my understanding, that's expected behavior; why would the compiler specialize `T` as `int` there? It could make sense to have a getter onto `Sealed<T>`, but then `consume` has no generic type associated and `value` becomes `this`.

There, I see Dart expects a type `T` to be returned, in a global function, whereas you're returning an `int`. It makes sense to have a compile-time error, there.

**Comment by lrhn** [comment]
@purplenoodlesoop It is indeed working as intended, as @lucavenir says. 
The `T` type variable is bound on method invocation, it's not updated or "promoted" during execution of the function body. That means that the type of `T` at the point of returning `1` is still just `T`, which, for all we statically know, could be some type which is not a supertype of `int`.

There is *currently* no subclasses of `Sealed` which has a type-argument other than `int`, but what if there was.
Say:
```dart
sealed class Sealed<T> {}

class Case implements Sealed<int> {}
class OtherCase implements Sealed<string> {}

T consume<T>(Sealed<T> value) => switch (value) {
      Case() => 1,
      OtherCase() => "wat",
};
```
Then we can't use a global analysis to decide that all actual objects implementing `Sealed` are `Sealed<int>`s.
It would have to use the `is Case` check to "promote" the type variable
But then it's also a matter of not having lower bounds in the type system in general.

Even if the type inference was incredibly clever, and recognized that any value which `is Case` must implement `Sealed<int>` (which they must, the only *current* subtype which doesn't is `Never`, which has no instances), all we'd know about `T` at that point is that it's a *supertype* of `int`. It might be `int`. It might be `Object`. It's definitely not `double` since all subclasses of `Case` must implement `Sealed<int>` precisely.

(I'm also not entirely convinced this doesn't hinge on `int` not having any unknown subtypes. Or that we don't have ways to create runtime-retained subtypes which are not subclasses, and which can therefore don't need to implement the same interfaces, just be a *subtype* of them. We can't currently, but future language features might be more flexible.)

Still, currently, knowning that a `Sealed<int>` is-a `Sealed<T>` is enough to conclude that `int` \<: `T` and allow returning `1` as `T`. 
The problem is then that the Dart type system equally currently doesn't have a way to represent "a type parameter which is some supertype of `int`". 


So complaining that `int` is not known to be assignable to `T` is correct here, there is no analysis, or currently possible analysis result, which would say that it is.

(If Dart gets variance annotations, it might also want to have lower bounds, like Java's `<X super int>`, to go with the contravariant type parameters. Or maybe that's more useful for use-side variance than declaration-side variance. If that happens, I still won't make an promises about what it will do to inference of this program. We'll just have to see what's practically possible.)
